src/go/configgenerator/filter_generator.go

```diff
@@ -0,0 +1,262 @@
+	commonpb "github.com/GoogleCloudPlatform/esp-v2/src/go/proto/api/envoy/v9/http/common"
+
+	routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	transcoderpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/grpc_json_transcoder/v3"
+	hcpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/health_check/v3"
+	routerpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/router/v3"
+	hcmpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	wrapperspb "github.com/golang/protobuf/ptypes/wrappers"
+	smpb "google.golang.org/genproto/googleapis/api/servicemanagement/v1"
+)
+
+// A wrapper of filter generation logic.
+type FilterGenerator struct {
+	// The filter name.
+	FilterName string
+	// The function to generate filter config and the methods requiring per route configs.
+	FilterGenFunc
+	// The function to generate per route config.
+	// It should be set if the filter needs to set per route config.
+	ci.PerRouteConfigGenFunc
+}
+
+// The function type to generate filter config.
+// Return
+//  - the filter config
+//  - the methods needed to add per route config
+//  - the error
+type FilterGenFunc func(sc *ci.ServiceInfo) (*hcmpb.HttpFilter, []*ci.MethodInfo, error)
+
+// MakeFilterGenerators provide of a slice of FilterGenerator in sequence.
+func MakeFilterGenerators(serviceInfo *ci.ServiceInfo) ([]*FilterGenerator, error) {
+	filterGenerators := []*FilterGenerator{}
+
+	if serviceInfo.Options.CorsPreset == "basic" || serviceInfo.Options.CorsPreset == "cors_with_regex" {
+		filterGenerators = append(filterGenerators, &FilterGenerator{
+			FilterName: util.CORS,
+			FilterGenFunc: func(sc *ci.ServiceInfo) (*hcmpb.HttpFilter, []*ci.MethodInfo, error) {
+				corsFilter := &hcmpb.HttpFilter{
+					Name: util.CORS,
+				}
+				return corsFilter, nil, nil
+			},
+		})
+	}
+
+	// Add Health Check filter if needed. It must behind Path Matcher filter, since Service Control
+	// filter needs to get the corresponding rule for health check calls, in order to skip Report
+	if serviceInfo.Options.Healthz != "" {
+		filterGenerators = append(filterGenerators, &FilterGenerator{
+			FilterName: util.HealthCheck,
+			FilterGenFunc: func(sc *ci.ServiceInfo) (*hcmpb.HttpFilter, []*ci.MethodInfo, error) {
+				hcFilter, err := makeHealthCheckFilter(serviceInfo)
+				if err != nil {
+					return nil, nil, err
+				}
+				return hcFilter, nil, nil
+			},
+		})
+	}
+
+	// Add JWT Authn filter if needed.
+	if !serviceInfo.Options.SkipJwtAuthnFilter {
+		// TODO(b/176432170): Handle errors here, prevent startup.
+		filterGenerators = append(filterGenerators, &FilterGenerator{
+			FilterName:            util.JwtAuthn,
+			FilterGenFunc:         jaFilterGenFunc,
+			PerRouteConfigGenFunc: jaPerRouteFilterConfigGen,
+		})
+	}
+
+	// Add Service Control filter if needed.
+	if !serviceInfo.Options.SkipServiceControlFilter {
```

Comment:
I am wondering if we can move such "SkipServiceControlFilter" check into each filterGen function.  It is filter specific.

If we move this into each filter, then each filter can register a factory (just like envoy factory register). At the high level,  we just loop over registered gen function to create create listener.

By doing this way,  OSS and g3 just need to have its own filter_gen folders.  just like envoy extensions.

Reply:
I think putting this logic here is OK. Basically `MakeFilterGenerators` will decide which filters to create. In OSS, it will decide based on the config generator flags/options. In g3, we will have an override of `MakeFilterGenerators` that will create filters differently.

It would be nice to have `Skip_${FilterName}_Filter` flags per filter, but honestly I don't think we will use them. It seems like an antipattern to have these flags for testing. The service config should decide which filters are generated, not random flags we pass in.

Reply:
I am not suggesting each filter should have a skip_xyz_filter flag.  I am suggesting "if" checking can be moved into each filter config_gen function.